src/main/java/com/ramcel/cinema/reservation/functionalities/ticket/validators/TicketValidator.java

```diff
@@ -0,0 +1,184 @@
+        }
+        return bookedSeatsByRowMap;
+    }
+
+    private SeatEntity fetchSeatEntity(Ticket ticket) throws IllegalSeatException {
+        return seatRepository.findById(ticket.getSeatId())
+                .orElseThrow(IllegalSeatException::new);
+    }
+
+    private static void checkEntityTicketConsistency(Ticket ticket, SeatEntity currentSeat) throws IllegalSeatException{
+        if(currentSeat.isOccupied() || haveNonMatchingScreenings(ticket, currentSeat)){
+            throw new IllegalSeatException("Seat can't be reserved - ticket data not in line with db.");
+        }
+    }
+
+    private static boolean haveNonMatchingScreenings(Ticket ticket, SeatEntity currentSeat) {
+        return currentSeat.getScreening().getId() != ticket.getScreeningId();
+    }
+
+    private void addRowToMap(Ticket ticket, SeatEntity currentSeat, Map<Long, List<SeatEntity>> bookedSeatsByRowMap) {
+        Long rowId = currentSeat.getRoomRow().getId();
+        Long currentScreening = ticket.getScreeningId();
+        List<SeatEntity> seatsInRow = getSeatsInRow(rowId, currentScreening);
+
+        bookedSeatsByRowMap.computeIfAbsent(rowId, v ->  seatsInRow);
+    }
+
+    private List<SeatEntity> getSeatsInRow(Long rowId, Long currentScreening) {
+        return new ArrayList<>(seatRepository.findSeatsByRoomRowIdAndScreeningID(rowId, currentScreening));
+    }
+
+    private static void reserveSeatsFromList(List<Ticket> ticketList, Map<Long, List<SeatEntity>> seatsByRowMap) throws IllegalTicketException{
+        Set<Long> seatsToBeBookedIds = ticketList.stream().map(Ticket::getSeatId).collect(Collectors.toSet());
+
+        for(List<SeatEntity> seatsInRow: seatsByRowMap.values()){
+            seatsInRow.stream().filter(v -> seatsToBeBookedIds.contains(v.getId())).forEach(SeatEntity::reserve);
+        }
+    }
+
+
+    private static boolean canBookSeatWithoutHoles(Map<Long, List<SeatEntity>> bookedSeatsByRowMap) {
+        return bookedSeatsByRowMap.values()
+                .stream()
+                .map(seats -> seats.stream()
+                        .sorted(Comparator.comparingInt(SeatEntity::getSeatNumber))
+                        .collect(Collectors.toList()))
+                .anyMatch(TicketValidator::hasNoHoles);
+    }
+
+    private static boolean hasNoHoles(List<SeatEntity> seats) {
+        if(seats.size() == 3){
+            if(hasAHole(seats)){
+                return false;
+            }
+        }
+        for (int i = 0; i < seats.size() - 2; i++) {
+            SeatEntity seat1 = seats.get(i);
+            SeatEntity seat2 = seats.get(i + 1);
+            SeatEntity seat3 = seats.get(i + 2);
+
+            if (seat1.getStatus() != SeatStatus.AVAILABLE
+                    && seat2.getStatus() == SeatStatus.AVAILABLE
+                    && seat3.getStatus() != SeatStatus.AVAILABLE) {
+                return false;
+            }
+        }
+        return true;
```

Comment:
Czy można byłoby to zapisać bez użycia `for`? :thinking:

Reply:
Dałoby się to zapisać jakimś functional chainem, ale nie próbowałem być zbyt cwany - for był naturalnym rozwiązaniem dla takiego algorytmiku. Problemem zwłaszcza standardowych streamów w javie byłoby tutaj to, że zapominamy o poprzednim elemencie i ciężko byłoby w naturalny sposób porównać ze sobą 3 elementy. Da się, tylko chyba nie warto.

Reply:
Cześć, pomyślałem jeszcze nad tym podpunktem chwile od wczoraj, bo stwierdziłem, że skoro pytasz, to jest jakiś sprytny myk jak to zrobić. Wymyśliłem coś takiego - do funkcji przekazujemy już listę jednego rzędu siedzeń, więc możemy posortować względem miejsca w rzędzie, odfiltrować puste miejsca i na koniec dać reduce, z warunkiem, że jeśli seatNr(i) = seatNr(i+1) + 2, to rzucamy false - mamy dziurę. To działa, ale dla mnie jest chyba trochę zbyt skomplikowane w porównaniu do fora. Co o tym myślisz? 
Jesli masz jakiś pomysł, jak to rozwiązać w iny sposób, to chętnie go poznam :)